exchange/utils.go

```diff
@@ -445,6 +453,100 @@ func copyWithBuyerUID(user *openrtb.User, buyerUID string) *openrtb.User {
 	return user
 }
 
+// removeUnpermissionedEids modifies the request to remove any request.user.ext.eids not permissions for the specific bidder
+func removeUnpermissionedEids(request *openrtb.BidRequest, bidder string, requestExt *openrtb_ext.ExtRequest) error {
+	// ensure request might have eids (as much as we can check before unmarshalling)
+	if request.User == nil || len(request.User.Ext) == 0 {
+		return nil
+	}
+
+	// ensure request has eid permissions to enforce
+	if requestExt == nil || requestExt.Prebid.Data == nil || len(requestExt.Prebid.Data.EidPermissions) == 0 {
+		return nil
+	}
+
+	// low level unmarshal to preserve other request.user.ext values. prebid server is non-destructive.
+	var userExt map[string]json.RawMessage
+	if err := json.Unmarshal(request.User.Ext, &userExt); err != nil {
+		return err
+	}
+
+	eidsRaw, eidsSpecified := userExt["eids"]
+	if !eidsSpecified {
+		return nil
+	}
+
+	var eids []openrtb_ext.ExtUserEid
+	if err := json.Unmarshal(eidsRaw, &eids); err != nil {
```

Comment:
Given that `extractBuyerUIDs` unmarshals the `User.Ext` into an `openrtb_ext.ExtUser` struct that already reads the `eids` into an `[]ExtUserEid` array, maybe we can work with this field and avoid additional unmarshalling inside `removeUnpermissionedEids`. Both `extractBuyerUIDs` and `removeUnpermissionedEids` use the unmarshalled `User.Ext`, therefore we could maybe:
```
174   func getAuctionBidderRequests(req AuctionRequest, requestExt *openrtb_ext.ExtRequest, impsByBidder map[string][]openrtb.I
175
176       bidderRequests := make([]BidderRequest, 0, len(impsByBidder))
177
    +     userExt := unmarshalUserExt(req.BidRequest.User)
    +
178 -     explicitBuyerUIDs, err := extractBuyerUIDs(req.BidRequest.User)
    +     explicitBuyerUIDs, err := extractBuyerUIDs(userExt)
179       if err != nil {
180           return nil, []error{err}
181       }
182
183   *-- 11 lines: var sChainsByBidder map[string]*openrtb_ext.ExtRequestPrebidSChainSChain---
194
195       var errs []error
196       for bidder, imps := range impsByBidder {
197           coreBidder := resolveBidder(bidder, aliases)
198
199   *--  5 lines: reqCopy := *req.BidRequest-------------------------------------------------
204
205           if err := removeUnpermissionedEids(&reqCopy, bidder, requestExt); err != nil {
    +         if err := removeUnpermissionedEids(&reqCopy, bidder, requestExt, userExt); err != nil {
206               errs = append(errs, fmt.Errorf("unable to enforce request.ext.prebid.data.eidpermissions because %v", err))
207               continue
208           }
exchange/utils.go
```
and modify `removeUnpermissionedEids` accordingly.